aiida_quantumespresso/parsers/ph.py

```diff
@@ -16,44 +18,85 @@ class PhParser(Parser):
     def parse(self, **kwargs):
         """Parse the retrieved files from a `PhCalculation`."""
         try:
-            output_folder = self.retrieved
+            self.retrieved
         except exceptions.NotExistent:
-            return self.exit_codes.ERROR_NO_RETRIEVED_FOLDER
-
-        list_of_files = output_folder.list_object_names()
+            return self.exit(self.exit_codes.ERROR_NO_RETRIEVED_FOLDER)
 
         # The stdout is required for parsing
         filename_stdout = self.node.get_attribute('output_filename')
         filename_tensor = PhCalculation._OUTPUT_XML_TENSOR_FILE_NAME
         filepath_dynmat = PhCalculation._FOLDER_DYNAMICAL_MATRIX
 
-        if filename_stdout not in list_of_files:
-            self.logger.error("The standard output file '{}' was not found but is required".format(filename_stdout))
-            return self.exit_codes.ERROR_READING_OUTPUT_FILE
+        if filename_stdout not in self.retrieved.list_object_names():
+            return self.exit_codes.ERROR_OUTPUT_STDOUT_MISSING
 
-        xml_tensor_file = None
-        if filename_tensor in list_of_files:
-            xml_tensor_file = output_folder._repository._get_base_folder().get_abs_path(filename_tensor)
+        try:
+            stdout = self.retrieved.get_object_content(filename_stdout)
+        except IOError:
+            return self.exit_codes.ERROR_OUTPUT_STDOUT_READ
+
+        try:
+            tensor_file = self.retrieved.get_object_content(filename_tensor)
+        except IOError:
+            tensor_file = None
 
         # Look for dynamical matrices
-        dynmat_dir = output_folder._repository._get_base_folder().get_abs_path(filepath_dynmat)
-        dynamical_matrix_list = [
-            os.path.join(dynmat_dir, file)
-            for file in os.listdir(dynmat_dir)
-            if os.path.isfile(os.path.join(dynmat_dir, file)) and
-            file.startswith(os.path.split(PhCalculation._OUTPUT_DYNAMICAL_MATRIX_PREFIX)[1]) and
-            not file.endswith('.freq')
-        ]
-
-        # Sort according to the number at the end of the dynamical matrix files, when there is one
+        dynmat_files = []
+        dynmat_folder = PhCalculation._FOLDER_DYNAMICAL_MATRIX
+        dynmat_prefix = os.path.split(PhCalculation._OUTPUT_DYNAMICAL_MATRIX_PREFIX)[1]
+
+        for filename in sorted(self.retrieved.list_object_names(dynmat_folder)):
```

Comment:
Not sure how important this is, but if the filenames have a numerical suffix, this is not equivalent to the sorting that we did previously:
```
# New:
In [22]: sorted(['a-1','a-2','a-11'])
Out[22]: ['a-1', 'a-11', 'a-2']

# Old:
In [23]: sorted(['a-1','a-2','a-11'], key=lambda x: int(x.split('-')[-1]))
Out[23]: ['a-1', 'a-2', 'a-11']
```

Reply:
Very good catch. This is indeed important because the parser that consumes it depends on the order. This is bad design in itself and should really be changed, but for now I will just fix the sorting here

Reply:
Ah, the tests now fail because we do the filtering after the sorting, and also we don't catch any exceptions from the lambda...